### PR TITLE
Tracker impl

### PR DESCRIPTION
**Pull request by GusevTimofey**

### CODE REVIEW

File: tracker/src/Tracker/Syntax/Retry.hs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import RIO
+import Control.Retry
+import Control.Monad.Catch
+import System.Logging.Hlog
+
+execWithRetry
+  :: (MonadIO f, MonadMask f)
+  => Logging f
+  -> String
+  -> f (a)
```
Comment (oskin1) [inline tracker/src/Tracker/Syntax/Retry.hs:12]:
```suggestion
  -> f a
```

File: tracker/src/Tracker/Services/Tracker.hs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  -> TrackerSettings
+  -> f ([CompletedTx], Int)
+getAllTransactions' Explorer{..} Cache{..} l@Logging{..} TrackerSettings{..} = do
+  _ <- infoM @String "Going to get next tracker iteration."
+  lastIndex <- getLastIndex
+  Items{..} <- execWithRetry l "Retrying getTxn." (getTxs (Paging lastIndex (naturalToInt limit)) Asc)
+  _ <- infoM $ "Got next txn batch: " ++ (show $ length items)
+  let 
+    cardanoTxn = fmap toCardanoTx items
+    newIndex = (length items) + lastIndex + 1
```
Comment (oskin1) [inline tracker/src/Tracker/Services/Tracker.hs:49]: Why not just tracking index by `Transaction.globalIndex` https://github.com/ergolabs/cardano-explorer-backend/blob/308052ed814d09f8ff793bd2ba64ba73f7324cec/modules/api/src/main/scala/io/ergolabs/cardano/explorer/api/v1/models/Transaction.scala#L22 ?

File: tracker/src/Tracker/Syntax/Option.hs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+module Tracker.Syntax.Option
+  ( unNone
+  ) where
+
+import RIO
+
+unNone :: [a] -> [Maybe a] -> [a]
```
Comment (oskin1) [inline tracker/src/Tracker/Syntax/Option.hs:7]: Why not make a function with a simpler signature and default acc=`[]`?

Comment (oskin1) [reply tracker/src/Tracker/Syntax/Option.hs:7]: `unNone :: [Maybe a] -> [a]`

File: tracker/src/Tracker/Caches/Cache.hs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  { getLastIndex :: f Int
+  , putLastIndex :: Int -> f ()
+  }
+
+mkCache
+  :: (MonadIO i, MonadResource i, MonadIO f) 
+  => RedisSettings
+  -> i (Cache f)
+mkCache settings =
+    fmap (\connection -> Cache  (getLastIndex' connection) (putLastIndex' connection)) connectionF
```
Comment (Bromel777) [inline tracker/src/Tracker/Caches/Cache.hs:25]: please add logging mid, for debug

File: tracker/src/Tracker/Models/AppConfig.hs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+instance FromDhall RedisSettings
+
+data TrackerSettings = TrackerSettings
+  { limit :: Natural
+  } deriving (Generic)
+
+instance FromDhall TrackerSettings
+
+data TrackerProgrammConfig = TrackerProgrammConfig
+   { pollTime :: Natural
```
Comment (Bromel777) [inline tracker/src/Tracker/Models/AppConfig.hs:23]: formatting

File: tracker/src/Tracker/Models/ExecutedOrders.hs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import CardanoTx.Models
+
+import           Data.Aeson (FromJSON, ToJSON)
+import qualified PlutusTx.Prelude as P    
+import           Ledger           
+import           GHC.Generics                (Generic)
+import           RIO
+
+data ExecutedSwap = ExecutedSwap
+  { config       :: Swap
```
Comment (Bromel777) [inline tracker/src/Tracker/Models/ExecutedOrders.hs:21]: you have the common pattern for all Executed(Swap/Deposit/Redeem): config, orderInputId, userOutputId, poolOutputId, poolInputId. Why not to create smth like ExecutedOrder a?

File: tracker/src/Tracker/Models/ExecutedOrders.hs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  , actualQuote  :: Amount Quote
+  , orderInputId :: String
```
Comment (Bromel777) [inline tracker/src/Tracker/Models/ExecutedOrders.hs:23]: Why not fullTxOutRef?

File: tracker/src/Tracker/Syntax/Retry.hs [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  -> f (a)
+execWithRetry Logging{..} logMsg func = do
+  recoverAll (constantDelay 1000000) (\rs -> (infoM $ logMsg ++ show rs) >> func)
```
Comment (Bromel777) [inline tracker/src/Tracker/Syntax/Retry.hs:15]: please, move `1000000` to cfg

File: tracker/src/Tracker/Syntax/Retry.hs (hunk 1 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Syntax/Retry.hs:12]: cleaned up

File: tracker/src/Tracker/Syntax/Option.hs (hunk 3 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Syntax/Option.hs:7]: simplified

File: tracker/src/Tracker/Models/AppConfig.hs (hunk 5 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Models/AppConfig.hs:23]: done

File: tracker/src/Tracker/Caches/Cache.hs (hunk 4 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Caches/Cache.hs:25]: done

File: tracker/src/Tracker/Syntax/Retry.hs (hunk 8 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Syntax/Retry.hs:15]: done

File: tracker/src/Tracker/Models/ExecutedOrders.hs (hunk 6 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Models/ExecutedOrders.hs:21]: done

File: tracker/src/Tracker/Models/ExecutedOrders.hs (hunk 7 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Models/ExecutedOrders.hs:23]: done

File: tracker/src/Tracker/Services/Tracker.hs (hunk 2 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Services/Tracker.hs:49]: https://github.com/ergolabs/cardano-explorer-backend/blob/308052ed814d09f8ff793bd2ba64ba73f7324cec/modules/core/src/main/scala/io/ergolabs/cardano/explorer/core/db/sql/TransactionsSql.scala#L28
doesn't support gix ordering

Comment (GusevTimofey) [reply tracker/src/Tracker/Services/Tracker.hs:49]: For example if you call `https://testnet-api.quickblue.io/v1/transactions/?offset=10&limit=1&ordering=asc` the result of **globalIndex** is **4587211**

File: tracker/src/Tracker/Models/ExecutedOrders.hs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+data ExecutedRedeem = ExecutedRedeem
+  { order   :: ExecutedOrder Redeem
+  , rewardX :: AssetAmount X
+  , rewardY :: AssetAmount Y
+  } deriving (Show, Eq, Generic, FromJSON, ToJSON)
+
+instance FromExplorer SettledTx ExecutedRedeem where
+  parseFromExplorer SettledTx{..} = do
+    (OnChain redeemOut redeem@Redeem{..}) <- findMatchedInput inputs :: Maybe (OnChain Redeem)
```
Comment (oskin1) [inline tracker/src/Tracker/Models/ExecutedOrders.hs:90]: Type application looks much nicer

**Review by oskin1 — APPROVED** [review]

File: tracker/src/Tracker/Models/ExecutedOrders.hs (hunk 9 above)
Comment (GusevTimofey) [reply tracker/src/Tracker/Models/ExecutedOrders.hs]: done

**Pull request merged**